io/src/lib.rs

```diff
@@ -0,0 +1,506 @@
+        where
+            E: Into<Box<dyn std::error::Error + Send + Sync + 'static>>,
+        {
+            Self { kind, error: Some(error.into()) }
+        }
+        #[cfg(all(feature = "alloc", not(feature = "std")))]
+        pub fn new<E: sealed::IntoBoxDynDebug>(kind: ErrorKind, error: E) -> Error {
+            Self { kind, error: Some(error.into()) }
+        }
+
+        pub fn kind(&self) -> ErrorKind {
+            self.kind
+        }
+    }
+
+    impl From<ErrorKind> for Error {
+        fn from(kind: ErrorKind) -> Error {
+            Self {
+                kind,
+                #[cfg(any(feature = "std", feature = "alloc"))]
+                error: None,
+            }
+        }
+    }
+
+    impl Display for Error {
+        fn fmt(&self, fmt: &mut Formatter) -> core::result::Result<(), core::fmt::Error> {
+            fmt.write_fmt(format_args!("I/O Error: {}", self.kind.description()))?;
+            #[cfg(any(feature = "alloc", feature = "std"))]
+            if let Some(e) = &self.error {
+                fmt.write_fmt(format_args!(". {:?}", e))?;
+            }
+            Ok(())
+        }
+    }
+
+    #[cfg(feature = "std")]
+    impl std::error::Error for Error {
+        fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+            self.error.as_ref().and_then(|e| e.as_ref().source())
+        }
+        #[allow(deprecated)]
+        fn description(&self) -> &str {
+            match self.error.as_ref() {
+                Some(e) => e.description(),
+                None => self.kind.description(),
+            }
+        }
+        #[allow(deprecated)]
+        fn cause(&self) -> Option<&dyn std::error::Error> {
+            self.error.as_ref().and_then(|e| e.as_ref().cause())
+        }
+    }
+
+    impl Error {
+        #[cfg(feature = "std")]
+        pub fn get_ref(&self) -> Option<&(dyn std::error::Error + Send + Sync + 'static)> {
+            #[allow(clippy::option_as_ref_deref)] // Doesn't yet meet MSRV
+            self.error.as_ref().map(|e| &**e)
+        }
+        #[cfg(all(feature = "alloc", not(feature = "std")))]
+        pub fn get_ref(&self) -> Option<&(dyn Debug + Send + Sync + 'static)> {
+            #[allow(clippy::option_as_ref_deref)] // Doesn't yet meet MSRV
+            self.error.as_ref().map(|e| &**e)
+        }
+    }
+
+    #[cfg(feature = "std")]
+    impl From<std::io::Error> for Error {
+        fn from(o: std::io::Error) -> Error {
+            Self { kind: ErrorKind::from_std(o.kind()), error: o.into_inner() }
+        }
+    }
+
+    #[cfg(feature = "std")]
+    impl From<Error> for std::io::Error {
+        fn from(o: Error) -> std::io::Error {
+            if let Some(err) = o.error {
+                std::io::Error::new(o.kind.to_std(), err)
+            } else {
+                o.kind.to_std().into()
+            }
+        }
+    }
+
+    #[cfg(all(feature = "core2", not(feature = "std")))]
+    impl From<core2::io::Error> for Error {
```

Comment:
I believe this is also non-additive because of what `core2` does (well, you could argue it's a mistake of `core2`, not this crate but that's beside the point).

Crate A depends on `core2`, crate B depends on `bitcoin-io`, crate C depends on A and `bitcoin-io`. At later point crate A starts depending on `core2/std`, this crate doesn't know it and `core2::io::Error` becomes the same type as `std::io::Error` and this becomes conflicting implementation.

Honestly, this whole business with `core2` being replaced with `std` is super annoying. I've been banging my head against the wall when writing `lgio` and concluded it's so broken that it's not worth trying to support. It's quite likely there are more subtle bugs in this code, so I suggest dropping the compatibility completely.